[1단계 - 지하철 정보 관리 기능] 홍실(홍혁준) 미션 제출합니다.

hong-sile opened the pull request:
안녕하세요! 범블비 리뷰이인 홍실입니다!!
이번 미션 잘 부탁드려요 😄

전체적인 구조에 대해서 설명드리겠습니다.

## 구조
Line은 Section을 구간으로 지니고 있습니다.
Section은 하나의 구간 단위를 의미하며, `상행 -> 하행` 방향으로 beforeStation과 nextStation을 필드로 지닙니다.

README에 적혀있는 API 명세외의 코드는 기본적으로 주어진 뼈대 코드라, API 명세서를 작성하지 않았습니다.

## 질문
1. 처음으로 단순한 CRUD가 아닌 도메인 로직과 웹요청이 섞인 형태의 어플리케이션 코드를 작성했습니다. 
어떤 흐름으로 짤 지 페어와 고민하다가, 다음과 같은 플로우로 코드를 작성하였는데, 아래와 같은 방식이 옳은지 궁금합니다.

현재 로직 수행 순서는 다음과 같습니다.
클라이언트의 요청 -> 
service에서 요청에 필요한 객체를 DB에서 조회 -> 
조회한 객체와 요청을 이용해서 도메인 로직 수행 ->
도메인 로직 수행 후 변한 값들만 찾아서, 디비에 반영 

위와 같은 방식으로 최대한 도메인 로직으로만 요구사항을 수행하고, 디비에 반영하려고 노력했습니다. 이런 흐름으로 코드를 작성하는 것이 맞을까요?

2. 현재는 없는 문제지만, 이전 코드는 DB의 구조와 도메인의 구조가 동일하지 않았습니다. 이때 DB 구조와 도메인 구조간의 괴리를 해결하기 위해 아래와 같은 방법을 시도했는데, 괜찮은 방법일지 궁금합니다.

Service <-> Repository 이 두 계층 간의 소통은 Domain 객체로 수행합니다.
Repository라는 객체는 Domain 객체를 받아서 저장하고, Domain 객체를 반환하는 역할을 해야 한다고 생각해서 위와 같은 구조를 짜봤습니다.

Repository <-> Dao 이 두 계층 간의 소통은 Entity 객체로 수행합니다. 
Entity는 DB구조와 일치하는 객체입니다.
Dao는 DB와 직접적으로 소통해서 저장 또는 조회를 합니다. DB를 조회한 값을 Entity로 받아서, 이를 Repository로 반환합니다.

이렇게 Repository에서는 Doamin객체를 받고 반환하고, 받은 Domain을 Entity로 변환해서 DB에 명령을 요청합니다.
이런식으로 작성하다가 중간의 구조가 바뀌고, 도메인과 디비구조가 일치해서 Repository 추상화 계층을 삭제하였습니다.
만약 위와 같은 방식으로 코드를 작성했다면 별 문제 없이 DB와 도메인 간의 괴리가 해소됬을까요?

로직의 디테일한 부분은 코드에 코멘트로 남기겠습니다.

(아 그리고, 구현하는데 시간이 모자라서 테스트가 좀 비어있긴 합니다... 차차 채워넣겠습니다.)

hyeonjerry committed:
chore: 제외 파일 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
docs: 요구사항 작성

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: 스키마 작성

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: line에서 color 필드 제거

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: station 테이블에 ID 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
style: Station 리팩터링

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: URI 수정

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: 더티 컨텍스트 제거

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: station의 pk를 id로 수정

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: LINE_NODE 테이블에 LINE_ID 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
test: 통합테스트 작성

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: 거리를 나타내는 Distance Vo 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: Node대신 Section으로 Line구성

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: Section 일급 컬렉션 구현

- 상행종점 여부 확인
- 하행종점 여부 확인
- 상행종점 Section 추가
- 하행종점 Section 추가
- 중간 Section 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: 노선에 역 추가 기능 구현

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: 노선에 역 제거 기능 구현

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
docs: 반영사항 체크

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
test: 역이 두개만 있을 때 하나를 제거하는 테스트 케이스 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: 라인 불변 객체로 변경

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: 두 Sections 간의 차집합을 구하는 기능 구현

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: Section 추가 기능 구현

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
chore: 테스트 환경 구성 및 데이터 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: Line의 getName이 노선명을 반환하도록 리팩터링

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: Line 조회 기능 구현

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: 메서드 분리

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
test: 역 등록 테스트 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: URI 수정 및 패키지 이동

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: 노선에 역을 등록하는 Api 구현

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: 노선에 역을 삭제하는 Api 구현

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: 노선에 역을 조회하는 Api 구현

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: 전체 노선의 목록을 조회하는 Api 구현

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: 미사용 dto 제거

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: 테스트 리팩터링

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
feat: request의 검증 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
docs: requestBody 에 대한 명세 추가

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
style: 컨벤션에 맞게 코드 정렬

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: 리스트 구현체 Linked List로 교체

Co-authored-by: hong-sile <gurwns9325@naver.com>

hyeonjerry committed:
refactor: 미사용 메서드 제거 및 메서드 추출

hong-sile committed:
chore: 패키지 변경

hong-sile committed:
style: 컨벤션에 맞게 코드 리포맷팅

hong-sile committed:
refactor: join으로 Section을 찾아오도록 리팩터링

hong-sile committed:
fix: 조회한 Line을 정렬하도록 리팩터링

hong-sile committed:
refactor: SectionService가 List<Section>을 반환하도록 리팩터링

hong-sile committed:
feat: Section의 포함된 Station이 동일한지 확인하는 기능 추가

hong-sile committed:
feat: section을 생성할 때 station들이 같은 경우 예외처리

hong-sile committed:
feat: 노선에 역을 추가할 때 이미 등록되어있는 section이면 예외처리한다.

hong-sile committed:
refactor: 미사용하는 생성자 삭제

hong-sile committed:
feat: 예외가 발생하면 예외메시지를 클라이언트에게 반환하는 기능 추가

hong-sile committed:
fix: 같은 station들을 지니고 있는지 확인하는 기능 수정

hong-sile committed:
feat: 디비에 값을 변경시키는 service 메서드에 Trasnactional 어노테이션 추가

hong-sile commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,48 +1,80 @@
+        }
+        final Sections addedSections = sections.addCentral(newSection);
+        return new Line(name, addedSections);
+    }
```
line에 값을 추가하거나, 삭제할 때, 기존의 값은 불변으로 두고 변한 값들을 반환하도록 코드를 작성했습니다.

hong-sile commented on src/main/java/subway/service/LineService.java:
```diff
@@ -0,0 +1,106 @@
+
+        final Sections deleteSections = line.getSections().getDifferenceOfSet(addedLine.getSections());
+        final Sections insertSections = addedLine.getSections().getDifferenceOfSet(line.getSections());
+
```
로직을 수행한 후의 Line과 로직 수행 전의 Line의 Sections에 대한 차집합을 구해서,
(변환 전의 Line) - (변환 후의 Line)의 차집합은 삭제해야할 레코드기에 delete요청
(변환 후의 Line) - (변환 전의 Line)의 차집합은 저장해야할 레코드기에 insert요청을 수행했습니다.

ddaaac replied:
> 클라이언트의 요청 ->
service에서 요청에 필요한 객체를 DB에서 조회 ->
조회한 객체와 요청을 이용해서 도메인 로직 수행 ->
도메인 로직 수행 후 변한 값들만 찾아서, 디비에 반영

아마 마지막 부분 때문에 질문 주신 것 같네요.
조금 더 도메인과 db와의 간극을 줄이려고 한다면 해당 line의 sections를 모두 지우고 새로 insert해도 될 것 같긴 하네요. 이러면 diff를 구하는 마치 db만을 위해 존재하는 로직이 사라지긴 할거에요.
하지만 지금 방식도 괜찮아보여요. 앞서 말한 방식은 불필요한 db write가 너무 많기도 하고, diff 구하는 로직도 비교적 간단한 로직이라 잘 구현해주시지 않았나 싶네요.

ddaaac commented on src/main/java/subway/controller/LineController.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<LineResponse> createLine(@RequestBody final LineRequest lineRequest) {
```
`@Valid`를 놓치신 것 같네요 😄

hong-sile replied:
앗.. 이 부분은 미션 시작할 때 기본적으로 주어진 뼈대 코드라서 따로 코드를 추가하진 않았어요.
추가해보도록 할게요!!

ddaaac commented on src/test/java/subway/domain/LineTest.java:
```diff
@@ -0,0 +1,151 @@
+            final Line removedLine = newLine.removeStation(STATION_A);
+            final List<Section> sections = removedLine.getSections().getSections();
+            assertAll(
+                    () -> assertThat(sections)
+                            .extracting(Section::getBeforeStation)
+                            .extracting(Station::getName)
+                            .containsExactly("b"),
+                    () -> assertThat(sections)
+                            .extracting(Section::getNextStation)
+                            .extracting(Station::getName)
+                            .containsExactly("c"),
+                    () -> assertThat(sections)
+                            .extracting(Section::getDistance)
+                            .extracting(Distance::getValue)
+                            .containsExactly(10)
```
적당히 메서드 분리를 하면 가독성을 올릴 수 있을 것 같네요.

+) 말씀해주신대로 빠진 테스트들 추가해주세요~

++) 테스트 깨지는 부분이 있네요. 확인해주세요.

hong-sile replied:
지금 보니 가독성이 너무 떨어지네요 수정하였습니다!!

+) 빠진 테스트 추가했어요!!

++) 조회에서 두 개가 깨지네요... 테스트 격리가 잘 안된 것 같습니다. 수정했어요!!

ddaaac replied:
Sections 테스트에서 차집합을 구하지 못하는 경우에 대한 테스트도 있으면 좋을 것 같네요.

hong-sile replied:
차집합을 구하지 못하는 경우가 어떤 경우인지를 모르겠어요.

Sections가 동일한 경우에만 차집합이 없으니 해당 상황에 대한 테스트 코드를 작성해봤습니다.

ddaaac replied:
아 차집합 메서드가 아니라 중간 역 제거 메서드에 대한 코멘트였어요. 코멘트를 잘못 남겼네요 😅

ddaaac commented on src/main/java/subway/service/dto/LineRequest.java:
```diff
@@ -0,0 +1,21 @@
+        this.name = name;
+    }
+
+    private LineRequest() {
+        this(null);
+    }
```
이 생성자는 왜 필요한가요?

hong-sile replied:
기본생성자가 없는 경우 Request가 ObjectMapper에 의해 정상적으로 역직렬화가 되지 않습니다.


그래서 private 생성자를 정의하고 그 안에서 값을 null로 초기화하도록 하였습니다.

private 생성자 안에서 값을 null로 초기화하는 이유는 필드에 final을 유지하기 위함입니다.
어차피 ObjectMapper에 의해서 역직렬화 될 때 값이 정상적으로 주입된 형태로 변화되기 때문에,
```java
private LineRequest(){
    this(null)
}
```
과 같은 생성자를 정의하였습니다.

ddaaac replied:
기본적으로 역직렬화 과정에서 all-arg constructor가 존재하면 jackson은 이를 사용합니다. 그래서 기본 생성자가 필요없습니다. 
만약 예외가 발생한다면 아래 이유 때문일텐데요
- all-arg constructor의 arg 갯수가 1일 경우거나
- gradle 대신 intelij 빌드를 사용해서

일 것 같네요. 전자의 경우에는 arg가 하나일 경우 잭슨의 동작방식 때문이고, 후자는 아마 [버그](https://youtrack.jetbrains.com/issue/IDEA-318002)지 않을까 싶네요.
이 외의 경우(직렬화, all-arg cstr가 존재하는 경우 등등)에는 기본 생성자가 필요없습니다. 관습적으로 기본 생성자를 사용한 코드가 많이 생기는 것 같아 남겨봐요.

hong-sile replied:
말씀해주신 두 가지 이유 모두 맞습니다.
gradle 대신 Intellij 빌드를 사용했을 때 오류가 생기는 이유는 두 빌드할때 추가하는 옵션의 차이때문에 그런걸로 알고 있습니다.

저는 빌드 환경에 따라 오류가 발생할 여지가 있다면 두 상태 모두에서 문제가 안 생기는 방법을 사용해야 된다고 생각해요.(제가 실 코드에서 빌드를 Intellij로 하기도 하고요.)
그리고, 기본생성자를 private으로 정의한다고 하더라도 문제가 생길 여지가 없다고 생각해서 추가하긴 합니다.

혹시, 위와 같은 기본생성자를 추가함으로써 생기는 문제가 있을까요?

ddaaac replied:
큰 문제는 없습니다. 관습적으로 기본 생성자를 생성하는 경우가 많아서 남긴 코멘트에요.

-----

보통 실제 배포할 때는 gradle build를 사용하는데요, intelij 빌드의 결과가 gradle 빌드의 결과가 다르다면 intelij 빌드를 신뢰할 수 있을지 의문이 드네요. all arg 생성자만 있어도 역직렬화가 가능하게 하려고 param name 모듈이 존재하는데 이를 지원하지 않는게 올바른 방향인지도 잘 모르겠구요.

hong-sile replied:
실제 빌드할때는 gradle을 사용하는군요. 
말씀해주신 모듈에 대해서 찾아보고 적용해보도록 하겠습니다.

ddaaac commented on src/main/java/subway/controller/LineController.java:
```diff
@@ -0,0 +1,82 @@
+        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+    }
+
+    @ExceptionHandler(SQLException.class)
+    public ResponseEntity<Void> handleSQLException() {
+        return ResponseEntity.badRequest().build();
+    }
```
SQL예외만 잡아서 핸들링 해주시는 이유가 있을까요?

hong-sile replied:
뼈대코드에 그대로 남아있는 코드에요.
ControllerAdvice.class에서 IllegalArgumentException을 핸들링하고 있습니다.

재리뷰 요청 보낼때는 해당 코드를 없애야겠네요

ddaaac commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,48 +1,80 @@
-        if (o == null || getClass() != o.getClass()) return false;
-        Line line = (Line) o;
-        return Objects.equals(id, line.id) && Objects.equals(name, line.name) && Objects.equals(color, line.color);
+    public boolean equals(final Object o) {
```
equals/hascode를 override 해준 기준이 있을까요?

hong-sile replied:
이미 뼈대 코드에 존재하는 equals hashcode를 수정해줬어요.
id만으로 동일한 객체로 판단할 수 있게요.

실 코드의 동작에 영향을 주는 코드는 아니고, 기존에 작성되어있는 코드를 단순하게 더 의미에 맞게 리팩터링했습니다.

hong-sile replied:
++ 이전 equals와 hashcode를 유지했던 이유는 기본적으로 주어진 뼈대코드에 영향을 주는 코드라 판단하여 유지하였는데, 지금 다시보니 전혀 equals와 hashcode를 재정의할 필요가 없네요. 

제거 후 제가 필요한 곳에서만 equals와 hashcode를 재정의했습니다.

1. Station
명확한 식별자(id)를 지니고 있고, 제가 이번에 구현한 `노선에 역 등록,삭제, 노선과 노선 목록 조회`에서는 이미 영속화된  Station만 사용할 것이기 때문에 id를 기준으로 equals hashcode를 재정의해도 된다고 생각했습니다.
재정의 하면서 테스트코드도 훨씬 짜기 용이해졌고요.

2. Distance
Distance는 VO입니다. 거리라는 값을 표현해주고, validation해주기 위한 클래스죠. 단순하 값만 나타내기에, equals hashcode를 value 기준으로 재정의했습니다.

ddaaac replied:
식별자를 갖고 있는 어떤 도메인은 override를 하고, 어떤 도메인은 하지 않아서 궁금해서 남긴 코멘트였습니다.
앞으로도 명확한 기준을 갖고 남겨주시면 될 것 같네요 👍

ddaaac commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,131 @@
+        return new Sections(newSections);
+    }
+
+    private static Section findOriginSection(final Section section, final LinkedList<Section> newSections) {
+        return newSections.stream()
+                .filter(element -> element.getBeforeStation().equals(section.getBeforeStation()))
+                .findAny()
+                .orElseThrow(() -> new IllegalArgumentException("이전 역을 찾을 수 없습니다."));
+    }
```
before만 확인해도 괜찮을까요?

hong-sile replied:
음 해당 코드를 짤땐 무조건 요청이
prevStation이 기존 존재하는 역임을 가정하면서 짰었는데, 이는 문제가 될 수 있겠네요.

맞춰서 도메인 로직을 수정해보겠습니다.

hong-sile replied:
짜고 보니 굉장히 긴 코드가 나왔네요... 
해당 부분 때문에 리뷰요청을 미룰 수 없으니, 다른 부분에 대한 리뷰를 받고 리팩터링을 진행해보도록 하겠습니다.

ddaaac commented on src/main/java/subway/domain/Section.java:
```diff
@@ -0,0 +1,44 @@
+public class Section {
+
+    private Long id;
+    private final Station beforeStation;
+    private final Station nextStation;
```
before/after, prev/next 이렇게 묶으면 좋을 것 같네요.

hong-sile replied:
확실히, 이름이 딱 맞지 않네요. 수정하도록 할게요!!

ddaaac commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,131 @@
+
+    private final List<Section> sections;
+
+    public Sections(final List<Section> sections) {
+        this.sections = sections;
+    }
```
`new Sections(new Section(1, 2), new Section(3, 4))`와 같이 Sections가 생성되면 어떻게 될까요?

hong-sile replied:
말씀해주신 것처럼 유효하지 않는 Sections가 생길 것 같아요. 해당 생성자를 private으로 막고,
addSecotions로만 Section을 Sections에 추가할 수 있도록 리팩터링 하였습니다.

ddaaac replied:
생성자에 검증 로직을 두는 방법도 고려해볼 법 합니다.
이런 코드가 있으면 코드를 읽는 다른 개발자가 도메인에 대해서 빠르게 파악할 수 있다는 장점도 있거든요.

hong-sile replied:
확실히 검증로직을 두는 방법도 있겠네요. 다른 개발자가 보는 상황은 생각지도 못했어요.

기존의 add와 remove를 이용하면 필요없을 것 같다는 생각도 들긴 하는데, 이 부분에 대해서는 조금 고민해보도록 할게요.

ddaaac commented on src/main/java/subway/domain/Section.java:
```diff
@@ -0,0 +1,44 @@
+
+public class Section {
+
+    private Long id;
+    private final Station beforeStation;
+    private final Station nextStation;
+    private final Distance distance;
```
여기서 line 정보를 뺀 이유가 있을까요?

hong-sile replied:
실제 자바 코드 상에서는
Line안에 `Sections(List <Section>)`를 갖고 있는 구조기에, 
Section에서 line에 대한 정보가 필요없다고 생각했어요.

DB컬럼과 동일하게 맞추기 위해 추가할까? 도 생각해봤지만, 
추가한다면 순환참조 형식이 될 것 같았고, 너무 DB테이블에 의존적인 도메인이 된 것 같아서, 삭제하였습니다.

ddaaac replied:
개인적으로는 Line 객체를 생성할 때 모든 Section이 같은 line에 속해있는지 검사하는 로직이 필요하지 않나 싶어요.
Line 대신 Line id를 갖고 있다면 직접적인 순환참조도 생기지 않을 거구요.

ddaaac commented on src/main/java/subway/service/LineService.java:
```diff
@@ -0,0 +1,106 @@
+        sectionService.insertSections(line.getId(), insertSections);
+    }
+
+    Line findById(final Long id) {
```
default 접근 제어자를 사용해주신 이유가 있을까요?

hong-sile replied:
해당 서비스를 메서드를 controller에서는 사용을 막고, service에서는 사용할 수 있도록 하기 위해 default 접근제어자를 두었습니다.

이 부분은 LineRepository라는 클래스를 만들고, 해당 클래스에 책임을 분리할 예정입니다.

ddaaac reviewed (changes requested):
안녕하세요 홍실 👋 

어려운 미션인데도 도메인 부분을 깔끔하게 구현해주셨네요.

몇 가지 코멘트 남겼으니 확인해주세요~

ddaaac commented on src/main/java/subway/persistence/dao/LineDao.java:
```diff
@@ -0,0 +1,59 @@
+
+@Repository
+public class LineDao {
+
```
> 만약 위와 같은 방식으로 코드를 작성했다면 별 문제 없이 DB와 도메인 간의 괴리가 해소됬을까요?

db 구조와 도메인 구조가 다른 문제를 해결하기 위해 많은 이론들과 라이브러리들이 나왔는데요, 개인적으로는 말씀해주신 db 엔티티와 도메인 객체를 따로 두는 방식이 제일 확실하고 명시적으로 문제를 해결할 수 있는 방법이라고 생각하긴 합니다.
다만 이 방식의 경우에도 보일러플레이트 코드가 너무 많이 발생하는 점과, mapping 과정에 로직이 들어갈 수도 있다는 점 등등 단점이 있긴 해요. 가능하면 같은 객체로 해결해보시고, 도메인과 db의 간극이 많이 벌어진다 싶으면 나눠보셔도 좋을 것 같네요.

hong-sile committed:
feat: createLine에서 @Valid추가

hong-sile committed:
fix: section이 있는 Line 조회할 때 id가 null로 반환되는 버그 수정

hong-sile committed:
test: 공통적으로 사용되는 section과 station 분리

hong-sile committed:
test: 거리를 나타내는 VO인 distance 테스트 코드 추가

hong-sile committed:
refactor: beforeStation -> prevStation으로 네이밍 변경

hong-sile committed:
test: prevStation과 nextStation이 동일한지 확인하는 기능 추가

hong-sile committed:
test: sections의 테스트 코드 추가

- sections의 상행종점에 section을 추가하는 기능 테스트
- sections의 하행종점에 section을 추가하는 기능 테스트
- 입력값이 sections의 상행종점인지 확인하는 기능 테스트
- 입력값이 sections의 하행종점인지 확인하는 기능 테스트

hong-sile committed:
test: Sections 중간에 추가하는 경우에 대한 테스트 코드 추가

hong-sile committed:
test: Section에서 station을 제거하는 기능 추가

- 상행종점을 제거한다.
- 하행종점을 제거한다.
- 중간에 있는 역을 제거한다.

hong-sile committed:
refactor: extracting을 하나로 묶어 검증하도록 리팩터링

hong-sile committed:
test: 이미 등록된 구간에 대해서 입력이 들어오는 경우 예외처리

hong-sile committed:
refactor: Line이 삭제되면 해당 Line에 포함된 모든 line이 삭제되도록 리팩터링

hong-sile committed:
test: extracting으로 가독성 향상

hong-sile committed:
test: 노선의 이름 길이가 20자가 넘는 경우 예외처리

hong-sile committed:
test: SectionDao 테스트 코드 추가

- Section 추가하는 기능 테스트
- Section 삭제하는 기능 테스트
- Section을 LineId로 조회하는 기능 테스트

hong-sile committed:
refactor: 미사용 메서드 삭제

hong-sile committed:
style: 컨벤션에 맞게 코드 수정

hong-sile committed:
refactor: Sections에서 List<Section>을 받는 생성자 private으로 변경

hong-sile committed:
refactor: 통합테스트에서 더미데이터는 Integration-data.sql만을 의존하도록 리팩터링

hong-sile committed:
test: 노선에 역을 등록하는 경우 테스트 코드 작성

- 이미 존재하는 구간을 등록하는 경우
- 거리가 1보다 작은 경우
- prev와 next가 동일한 경우

hong-sile committed:
feat: List에 새 Station을 등록할 때, 이미 Station이 모두 포함되어있으면 exception 반환

hong-sile committed:
feat: 들어온 요청이 prev만 일치하는 케이스가 아니라 next도 일치하는 케이스도 적용되도록 수정

hong-sile committed:
refactor: 무의미한 Line의 equals hashcode 제거

hong-sile committed:
feat: Line에 존재하지 않는 역을 삭제하려는 경우 예외처리

ddaaac commented on src/main/java/subway/controller/ControllerAdvice.java:
```diff
@@ -0,0 +1,14 @@
+@RestControllerAdvice
+public class ControllerAdvice {
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<String> handleException(final IllegalArgumentException exception) {
+        return ResponseEntity.badRequest().body(exception.getMessage());
+    }
```
- 정의하지 않은 예외가 발생하면 어떻게 될까요?
- 라이브러리에서 IllegalArgumentException을 터트릴 경우 클라이언트에서는 어떤 일이 일어날 수 있을까요?

hong-sile replied:
- InternalServerError 500이 클라이언트에게 반환될 것 같아요.
RunTimeException을 핸들링해서, 로그를 찍는 로직을 추가해볼게요.

- 라이브러리에서 발생한 에러가 직접적으로 클라이언트에게 보여질 것 같네요.
CustomException을 정의해서, 핸들링해보도록 하겠습니다.

ddaaac commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,181 @@
+
+public class Sections {
+
+    private static final Map<BiPredicate<Section, Section>, TriConsumer<List<Section>, Section, Section>> addPrincipal = Map.of(
```
이 친구는 어디에 쓰일까요?
함수형을 사용하는 게 가독성에 도움이 될 때도 있지만, 이런 경우도 가독성에 도움이 될지 생각해보시면 좋을 것 같네요.

hong-sile replied:
아 추상화를 어떻게 잘해볼까 추가해봤다가 미쳐 지우지 못한 코드네요.
바로 지웠습니다.

ddaaac commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,181 @@
+        final Section originSection = findOriginSection(section, newSections);
+
+        final int originIndex = newSections.indexOf(originSection);
+        if (originSection.getNextStation().equals(section.getNextStation())) {
+            newSections.add(originIndex, section);
+            newSections.add(originIndex,
+                    new Section(
+                            originSection.getPrevStation(),
+                            section.getPrevStation(),
+                            originSection.getDistance().minusValue(section.getDistance())
+                    )
+            );
+        }
+        if (originSection.getPrevStation().equals(section.getPrevStation())) {
+            newSections.add(originIndex,
+                    new Section(
+                            section.getNextStation(),
+                            originSection.getNextStation(),
+                            originSection.getDistance().minusValue(section.getDistance())
+                    )
+            );
+            newSections.add(originIndex, section);
+        }
```
이 부분 때문에 메서드가 길어지네요.
section을 다른 section으로 자르는 메서드를 Section에 추가해도 좋을 것 같아요.

hong-sile replied:
아 말씀해주신 로직을 Section에 주면, 공통 관심사항을 분리할 수 있을 것 같네요.
한번 시도해보겠습니다.

ddaaac commented on src/main/java/subway/domain/Sections.java:
```diff
@@ -0,0 +1,181 @@
+                    new Section(
+                            originSection.getPrevStation(),
+                            section.getPrevStation(),
+                            originSection.getDistance().minusValue(section.getDistance())
```
이 부분에서 예외가 터질 경우에 클라이언트가 보게 될 에러 메세지가 적당한지에 대해서도 생각해보시면 좋을 것 같습니다.

hong-sile replied:
아마 minusValue가 터졌을 때, 나오는 exception을 말씀해주신 것 같네요.
이 부분이 클라이언트가 봤을 때 이해가 안 되는 상황일 것 같긴합니다.

다시 catch해서, 새로운 Exception을 던지는 방향으로 리팩터링 해볼게요.

ddaaac reviewed (approved):
안녕하세요 홍실 👋 

리뷰 반영 잘 해주셨습니다~

몇 가지 코멘트를 남겼으니 다음 단계에서 반영해주세요!
